### ISSUE TITLE
Softmax for Image Segmentation

### ISSUE THREAD

**Issue by meanmee** [Theano/Theano#4269]
I am trying  to implement the Fully Convolutional Networks via Keras and Theano. But When I add Softmax after the Deconvolution Layer， There an error raise: softmax do nor support 4-D input.

So, I just wonder any easy way to implement the pixel-wise softmax? not just a for loop.

**Comment by benanne** [comment]
You can fold the spatial dimensions into the batch dimension using a dimshuffle and a reshape: basically you go from (batch_size, channels, height, width) to (batch_size * height * width, channels), then apply the softmax on this 2D array, and then reshape the result back to the original shape.

We've been discussing this use case for Lasagne ( https://github.com/Lasagne/Lasagne/issues/626 ), but I suppose it might be useful to have support for this in Theano instead.

**Comment by nouiz** [comment]
We could create something for that in Theano. Do you have a suggestion for
the interface?

**Comment by benanne** [comment]
As Jan pointed out in the Lasagne discussion thread, we would need special interfaces for both the softmax and the categorical cross-entropy. The 2D case will obviously be the most common, but it would be nice to transparently support 1D and 3D (and beyond?) as well. Finally, I think it is important that the API clearly distinguishes the softmax-over-locations and the softmax-per-location, since both are quite common nowadays, so coming up with a good name that allows for that is instrumental.
